src/import/engine.rs

```diff
@@ -0,0 +1,172 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp;
+use std::i32;
+use std::fmt;
+use std::ops::Deref;
+use std::sync::Arc;
+use std::path::Path;
+
+use uuid::Uuid;
+
+use rocksdb::{BlockBasedOptions, ColumnFamilyOptions, DBOptions, Writable, WriteBatch as RawBatch,
+              DB};
+use kvproto::importpb::*;
+
+use config::DbConfig;
+use storage::CF_DEFAULT;
+use storage::types::Key;
+use util::config::MB;
+use util::rocksdb::{new_engine_opt, CFOptions};
+use util::rocksdb::properties::SizePropertiesCollectorFactory;
+
+use super::Result;
+
+pub struct Engine {
+    db: Arc<DB>,
+    uuid: Uuid,
+}
+
+impl Engine {
+    pub fn new<P: AsRef<Path>>(path: P, uuid: Uuid, opts: DbConfig) -> Result<Engine> {
+        let db = {
+            let (db_opts, cf_opts) = tune_dboptions_for_bulk_load(&opts);
+            new_engine_opt(path.as_ref().to_str().unwrap(), db_opts, vec![cf_opts])?
+        };
+        Ok(Engine {
+            db: Arc::new(db),
+            uuid: uuid,
+        })
+    }
+
+    pub fn uuid(&self) -> Uuid {
+        self.uuid
+    }
+
+    pub fn write(&self, mut batch: WriteBatch) -> Result<usize> {
+        // Just a guess.
+        let wb_cap = cmp::min(batch.get_mutations().len() * 128, MB as usize);
+        let wb = RawBatch::with_capacity(wb_cap);
+        let commit_ts = batch.get_commit_ts();
+        for m in batch.take_mutations().iter_mut() {
+            let k = Key::from_raw(m.get_key()).append_ts(commit_ts);
```

Comment:
Could you reuse the `k`?

Reply:
What's that mean? The key is different every time.

Reply:
Reuse its alloction.

Reply:
It doesn't seem we have a way to reuse it now, we can optimize this in the future when necessary.